icefoganalytics/travel-authorization
# Update Per-Diem Logic Calculation for Estimate Generation for Travel Authorizations

Fixes https://github.com/icefoganalytics/travel-authorization/issues/121

Relates to:

- https://github.com/icefoganalytics/travel-authorization/issues/68

# Context

Original per-diem estimates where something I made up on the fly, and had no relation to actual requirements. As such, the _real_ logic needs to be implemented.

See [api/src/services/estimates/bulk-generate.ts](https://github.com/icefoganalytics/travel-authorization/blob/08a8492fdfdce61a301d15442c62a041d65dd8cb/api/src/services/estimates/bulk-generate.ts#L315)

**Store the logic in an obvious place, as this logic might be contentious amongst department, and may change. It needs to be easy to read, and well documented.** - probably just break this calculation out into a sub-service and write a bunch of tests for it.

## Requirements

1. On the first day of travel:
    1. Where travel starts after 8am the traveler will not receive breakfast.
    2. Where travel starts after 1pm the traveler will not receive breakfast or lunch.
    3. Where travel starts after 7pm the traveler will not receive any per diem.
        > these rules are going to be slightly different when we look at air travel as travelers need to be at the airport 2 hours early but for the estimates this will work as we are not clear on exact travel times
    4. No incidentals can be claimed this day.

4. On the second and subsequent days of travel:
    1. The traveler will receive the full per diem including incidentals.

5. On the last day of travel:
    1. Where travel ends at 10am the traveler will receive only breakfast
    2. Where travel ends at 2pm the traveler will receive breakfast and lunch
    3. Where travel ends at 7pm the traveler will receive breakfast, lunch and supper
    4. Incidental can be claimed this day regardless of arrival time.

# Implementation

1. Move per-diem seeds to a .csv file to make the code easier to read. Also lets user edit seed data via spreadsheet software, instead of having to look at the code.
2. Drop "maximum daily" per diem claim types, replace with "incidentals" claim type. This lets us correctly calculate per-diems for first/last day segments.
3. Move per-diem claim calculations to a separate function for easier testing/verification.
4. Calculate per-diems against the entire travel date range, instead of per-day; this lets us correctly calculate per-diems for first/last day segments.
5. Calculate per-diem claim location via independent function. The logic doesn't quite make sense, so we might be missing something in the data model, or the original user story might not be valid.

# Screenshots

![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/dc837186-c034-4797-9bcd-8cc3ac743522)
> Note that values are less than in the user story as per-diems are being calculated as occurring in the "rest of canada" i.e. BC where the travel is taking place.

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
7. Boot the app via `dev up`
8. Log in to the app at http://localhost:8080
9. Refresh the seeds by going to http://localhost:3000/migrate/seed. **Note that this wipes the all data in the database.** See open ticket at https://github.com/icefoganalytics/travel-authorization/issues/72
10. Go to the "My Travel Requests" page via the top drop down nav.
11. Create a new "Travel Authorization" via the top right button.
12. Go to the "Details" section.
13. Make a trip from Whitehorse to Vancouver on the 5th of the current month.
14. Don't set a time for the departure date from Whitehorse.
15. Leave the travel method and accommodation type as their defaults of Aircraft + Hotel.
16. Make the return date the 7th of the month at 15:00 (3pm).
17. Go to the "Approvals" Section.
18. Generate the estimate via the call-to-action button.
19. Click the "Edit Estimate" button to view the generated estimates.
20. Check that the estimates calculations for per-diems match expected values calculated in BC (Rest of Canada).
    1. Note that the first day gets counted as a full day because time was not provided.
    2. Note that the first day does not get "incidentals", but the second and sub-sequent days do.
    3. Note that the last day does not get any supper per-diem as the trip ended before 7pm.
